### ISSUE TITLE
Use a persistent client connection as a server

### ISSUE THREAD

**Issue by fdawg4l** [dustin-decker/grpc-firewall-bypass#1]
Thanks for posting this.  It did _nearly_ everything I needed.  The issue was `Accept()` in the client would pile up unused connections to the server, and this didn't work for my use-case.

So I used `yamux` to re-use a single connection.  It's hacky, but works really well.  Anyway, link to commit to follow.  Just wanted to let you know.

**Comment by fdawg4l** [comment]
https://github.com/fdawg4l/grpc-firewall-bypass/commit/0cc2f79050bb0e26c3c5ba59610fa564eabb6e9b

**Comment by dustin-decker** [comment]
Thanks for sharing. I'm glad it was useful for you.

Just so i'm clear, is there a bug with the connections not being closed by the TCP server (server/main.go) when the TCP client (client/main.go) dies? Or did you need to create multiple connection to the client's gRPC server (client/main.go) over the same connection?

**Comment by fdawg4l** [comment]
The unmodified code does this (https://github.com/dustin-decker/grpc-firewall-bypass/blob/master/client/main.go#L21).  The grpc server will do the standard `Accept()` loop thing, call the client's `listener.Accept()` and open a new connection to the `server`.  In effect, the grpc server will open as many file descriptors as it can before the OS runs out, then the server will reply to each one with an echo.  Once replied, the client `listener.Close()` is called, dropping the connection.  

Put a `time.Sleep(3 * time.Second)` on line 31 in `server/main.go`, and watch the client's terminal spew messages about connecting with TCP client.  Each one a connection to the server.

It's entirely possible that I'm completely wrong.  Apologies in advance if that's the case.  I wanted to avoid making more than 1 persistent connection to the server from the client.

**Comment by dustin-decker** [comment]
Ah, I see. I ran your changes, but found the TCP server could only receive the one connection because the *Listener.Accept() was not in a loop, which might be desired in your use case. Once I moved [lines 29-41](https://github.com/fdawg4l/grpc-firewall-bypass/blob/0cc2f79050bb0e26c3c5ba59610fa564eabb6e9b/server/main.go#L29-L41) of server/main.go back into the for loop so the TCP server can receive new connections, it does not repeat from redialing the connection like it did before, and is the behavior I originally intended. So thanks! That's a big improvement. Would you like to make a PR with it that way?

**Comment by dustin-decker** [comment]
I ran these under strace to verify behavior. Looks good with the changes:
```
➜  grpc-firewall-bypass git:(master) ✗ strace -f -e trace=network -s 10000  ./bin/server
strace: Process 18214 attached
strace: Process 18215 attached
strace: Process 18216 attached
strace: Process 18217 attached
strace: Process 18218 attached
2018/06/06 20:02:12 launching tcp server...
[pid 18213] socket(AF_INET, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_TCP) = 3
[pid 18213] socket(AF_INET6, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_TCP) = 3
[pid 18213] setsockopt(3, SOL_IPV6, IPV6_V6ONLY, [1], 4) = 0
[pid 18213] bind(3, {sa_family=AF_INET6, sin6_port=htons(0), inet_pton(AF_INET6, "::1", &sin6_addr), sin6_flowinfo=htonl(0), sin6_scope_id=0}, 28) = 0
[pid 18213] socket(AF_INET6, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_TCP) = 5
[pid 18213] setsockopt(5, SOL_IPV6, IPV6_V6ONLY, [0], 4) = 0
[pid 18213] bind(5, {sa_family=AF_INET6, sin6_port=htons(0), inet_pton(AF_INET6, "::ffff:127.0.0.1", &sin6_addr), sin6_flowinfo=htonl(0), sin6_scope_id=0}, 28) = 0
[pid 18213] socket(AF_INET6, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 3
[pid 18213] setsockopt(3, SOL_IPV6, IPV6_V6ONLY, [0], 4) = 0
[pid 18213] setsockopt(3, SOL_SOCKET, SO_BROADCAST, [1], 4) = 0
[pid 18213] setsockopt(3, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0
[pid 18213] bind(3, {sa_family=AF_INET6, sin6_port=htons(8081), inet_pton(AF_INET6, "::", &sin6_addr), sin6_flowinfo=htonl(0), sin6_scope_id=0}, 28) = 0
[pid 18213] listen(3, 128)              = 0
[pid 18213] getsockname(3, {sa_family=AF_INET6, sin6_port=htons(8081), inet_pton(AF_INET6, "::", &sin6_addr), sin6_flowinfo=htonl(0), sin6_scope_id=0}, [112->28]) = 0
2018/06/06 20:02:12 waiting for incoming TCP connections...
```

This is when it's waiting on a Listener.Accept():
```
[pid 18213] accept4(3, 0xc42006bbc0, [112], SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
```

Here's where I turn on a client to make a connection:
```
[pid 18213] accept4(3, {sa_family=AF_INET6, sin6_port=htons(52874), inet_pton(AF_INET6, "::ffff:127.0.0.1", &sin6_addr), sin6_flowinfo=htonl(0), sin6_scope_id=0}, [112->28], SOCK_CLOEXEC|SOCK_NONBLOCK) = 5
[pid 18213] getsockname(5, {sa_family=AF_INET6, sin6_port=htons(8081), inet_pton(AF_INET6, "::ffff:127.0.0.1", &sin6_addr), sin6_flowinfo=htonl(0), sin6_scope_id=0}, [112->28]) = 0
[pid 18213] setsockopt(5, SOL_TCP, TCP_NODELAY, [1], 4) = 0
2018/06/06 20:02:17 starting a gRPC server over incoming TCP connection
strace: Process 18232 attached
strace: Process 18234 attached
strace: Process 18235 attached
2018/06/06 20:02:17 response from server: bar
strace: Process 18237 attached
2018/06/06 20:02:20 waiting for incoming TCP connections...
[pid 18213] accept4(3, 0xc42006bbc0, [112], SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
```

**Comment by fdawg4l** [comment]
> back into the for loop so the TCP server can receive new connections,

Are you starting more than one client process?  As written, this is unnecessary because the client only connects once.  And, yes, this is exactly my use case.  But you are right-  to support more than 1 "server" (nee client), this needs to be in the loop.

I'll make the change you're suggesting and push a PR.  Thanks for looking into it and verifying my reading of the code.  Much appreciated!

**Comment by dustin-decker** [comment]
Yes, I am using multiple clients. My use case was for [threatseer](https://github.com/dustin-decker/threatseer), which has a bunch of agents (TCP client, and gRPC server) that connect to a server (TCP server, gRPC client that subscribes to telemetry). 
This workaround really has made it confusing to talk about clients and servers.

**Comment by fdawg4l** [comment]
I use "initiator" and "target" to refer to the muxed (top most) connections.

**Comment by dustin-decker** [comment]
Closed by #2
